src/Compilers/Core/Portable/SourceGeneration/RunResults.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.Text;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents the results of running a generation pass over a set of <see cref="ISourceGenerator"/>s.
+    /// </summary>
+    public class GeneratorDriverRunResult
+    {
+        private Lazy<ImmutableArray<Diagnostic>> _lazyDiagnostics;
+
+        private Lazy<ImmutableArray<SyntaxTree>> _lazyGeneratedTrees;
+
+        internal GeneratorDriverRunResult(ImmutableArray<GeneratorRunResult> results)
+        {
+            this.Results = results;
+            _lazyDiagnostics = new Lazy<ImmutableArray<Diagnostic>>(() => results.SelectMany(r => r.Diagnostics).ToImmutableArray());
+            _lazyGeneratedTrees = new Lazy<ImmutableArray<SyntaxTree>>(() => results.SelectMany(r => r.GeneratedSources.Select(g => g.SyntaxTree)).ToImmutableArray());
```

Comment:
Lazys have historically bitten us in allocations, particularly in things like this that will be allocated constantly. Consider using a backing field an `InterlockedImmutable.Initialize` instead.